### DISCUSSION TITLE
Slicing along axis in reverse order

### DISCUSSION THREAD

**Discussion by at88mph** [nom-tam-fits/nom-tam-fits#618] [Q&A]
We have a use case for converting FITS to JPEG, where the starting corners are different.  As such, we'd like to read the FITS file along one axis, but flip it along the second axis.

Currently, slicing options are done by setting the start positions (`corners`), and then setting the lengths along each dimension, which assumes reading in one direction.  Typically, to read 100 values along both axes, we would specify something like:

```java
final int[] corners = new int[]{0, 0};
final int[] lengths = new int[]{100, 100};
tiler.getTile(corners, lengths);
```

But what I'm requesting is something like:

```java
final int[] corners = new int[]{0, 100};
final int[] lengths = new int[]{100, -100};
tiler.getTile(corners, lengths);
```

Where the second example would start at a different (opposite) side and read backward.  I don't see a way that's possible now, but before I work on it I wanted to ask.

Best.

**Comment by attipaci** [comment]
I certainly would not object to having a feature like that, if you want to work on it. I'll give it some thought on how I would implement it, and then we can exchange notes -- hopefully to find a robust, elegant and generic way to do it... If you don't hear back from me in a couple of weeks, ping me again here... Or if you have ideas, you can start by outlining them here also...

**Comment by attipaci** [comment]
The slicing you describe is pretty easy for arrays. I have produced a generic implementation in less than 20 lines of code, which I can add to `ArrayFuncs`, and does what you described for arrays, e.g.

```java
  // Slicing method where lengths can be negative to go backwards...
  Object ArrayFuncs.slice(Object array, int[] corners, int[] lengths);
``` 

For slicing from other sources (e.g. streams), I'm thinking of a 2-step process:

1. You obtain the cutout as before in the same ordering, adjusting the corners as necessary for all positive lengths. E.g. for corner `{0, 100}`, and length `{100, -100}`, you produce the cutout for the adjusted corner `{0, 0}` and size `{100, 100}`.
2. Then, you get a flipped cutout for the axes where the original length was negative, in a seconds step.

Step 1 for slicing from non-array Objects (e.g. streams) will look like:

```java
  // Adjusted corners / lengths but for a regular slice with no reversed ordering...
  int[] from = Arrays.copyOf(corners, lengths.length);
  int[] sizes = Arrays.copyOf(lengths, lengths.length);

  // Whether any of the axes are to be reversed...
  boolean flipped = false;

  // Calculate adjusted corners and non-negative lengths for a 'regular' slice...
  for (int i = 0; i < lengths.length; i++) {
    if (lengths[i] < 0) {
      from[i] += lengths[i];
      sizes[i] = -lengths[i];
      flipped = true;
    }
  }

  // Now do the regular slice with from[] and sizes[]...
  Object slice = ...

  if (!flipped) {
    return slice;
  }
  
  // Adjust the corners for the normally ordered slice...
  for (int i = 0; i < lengths.length; i++) {
    from[i] = corners[i] - from[i]; 
  }

  return ArrayFuncs.slice(slice, from, lengths);
```

If you agree, I'll add the `ArrayFuncs` slicing methods to a branch and PR, and you can work with that...

How does that sound?

**Comment by at88mph** [comment]
I added the `step` feature some time ago as well, so the `ArrayFuncs.slice` should support that, I think.

Wouldn't the `ArrayFuncs.slice` be a more generic method in general?  The `StandardImageTiler`, for example, has very complicated looping that took me a while to figure out, but did you encompass that in that method?

I'm happy to work on it, though.  It's one of those odd features that one or two Astronomers requested...

**Reply by attipaci** [reply]
Hi Dustin,

OK, I have modified it a bit and added a family of `ArrayFuncs.sample()` methods as well as the above mentioned `ArrayFuncs.slice()` (the latter has no `step` argument -- it is always densely sampled). You can find them in PR #619.

I haven't really tested these yet. I'll add unit tests soon (and fix anything that pops up...)

Feel free to use it if it helps with what you are planning to implement (or don't if it's not convenient for what you are planning). I guess, you can either revise the existing (already complicated) `StandardImageTiler` sampling to account for negative sizes (or negative steps?), or else you can do what I suggested above and call one of the new `ArrayFuncs` methods on the result of the existing non-reversed sampling. (The former will likely increase code complexity a fair bit, but would be more efficient). Maybe in re-thinking / re-writing the tiling from the ground up might actually make it simpler. If so, that would be the best option for sure...

Some ideas for making things more user-friendly:

- allow `corners`, `lengths` and `steps` to be `null` -- causing them to be default (to start at zero offsets, sample the full array, with steps of 1 -- respectively). See the private `ArrayFuncs.sample(...)` for an example how...
- allow individual elements of `lengths` and `steps` to be 0 to indicate that a default should be used for these for an individual axis, similarly to to sample the full extent and/or with a step of 1, respectively.

Have fun with it. My guess is that we should not rush it into the 1.20 release (in a month). Rather, we can integrate it to master (SNAPSHOT) whenever it is ready, and then include it as an trial feature in the next update release (1.20.1) in September...

**Reply by attipaci** [reply]
The `ArrayFuncs` additions are now fully tested (after some tweaking and necessary fixes). That part should be good to go should you need it.

**Comment by at88mph** [comment]
Well done, @attipaci .  I will start work this week.

**Comment by at88mph** [comment]
Thanks for the contribution, @attipaci.  I'm adapting it this week.

I think we may be able to simplify some of the Tiler logic, or at least share some of the `ArrayFuncs` methods.

**Comment by at88mph** [comment]
How have you been managing branches and tags?  My target would just be `master` for the time being, I think.  Do you typically branch off closer to release?

**Reply by attipaci** [reply]
I think it's OK to branch `master`. It's just a question of whether we merge it before or after the upcoming release. If you can get the new feature ready with 100% test coverage to your (and my) full satisfaction by the end of the month, then we can certainly include it on the 1.20.0 release around June 15. Otherwise, we may just merge it to master after, and then release it in September. Either way, the important thing is to do it well, and get it right, so we don't have to back-pedal on the choices made later on...

**Comment by attipaci** [comment]
Hi @at88mph, 

I'm just checking in to see where you are with your new feature implementation. I'm planning on 1.20.0-rc2 next week (probably on Monday). After that I'd like to code freeze until the finalized 1.20.0 release, notwithstanding any last minute bug fixes.

I don't want you to rush it though -- it's better to take your time, and also for me to review it without pressure. We can integrate it to master any time after the release, and include it as a 'preview' feature in 1.20.1 in 3 months, with the final API set in stone for the 1.21 in December.

**Comment by at88mph** [comment]
Hi @attipaci ,

Looking a few Sprints ahead, It will likely be a part of the 1.21 release in December.  Thanks for checking in.
